### PR TITLE
Feature plane approximation utility

### PR DESCRIPTION
**Pull request by rubenzorrilla**

### CODE REVIEW

File: kratos/utilities/plane_approximation_utility.h [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+     * @param rPointsCoords Vector containing the set of point coordinates
+     * @return rPlaneBasePointCoords Plane base point
+     * @return rPlaneNormal Plane normal
+     */
+    static void ComputePlaneApproximation(
+        const std::vector< array_1d<double,3> > &rPointsCoords,
+        array_1d<double,3> &rPlaneBasePointCoords,
+        array_1d<double,3> &rPlaneNormal)
+    {
+        rPlaneBasePointCoords = GetPlaneBasePoint(rPointsCoords);
```
Comment (RiccardoRossi) [inline kratos/utilities/plane_approximation_utility.h]: it would be better to pass the output by argument, not to rely on the optimizer

**Review by RiccardoRossi — DISMISSED** [review]
i am approving, but it would be better to pass the parameter by argument instead of returning it...
(the optimizer will do the work for you, so i am accepting)

File: kratos/utilities/plane_approximation_utility.h (hunk 1 above)
Comment (rubenzorrilla) [reply kratos/utilities/plane_approximation_utility.h]: I'm changing it.

**Review by RiccardoRossi — APPROVED** [review]

File: kratos/utilities/plane_approximation_utility.h [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        array_1d<double,3> &rPlaneNormal) 
+    {
+        // Solve the A matrix eigenvalue problem 
+        bounded_matrix<double, 3, 3> a_mat, eigenval_mat, eigenvector_mat;
+        SetMatrixA(rPointsCoords, rPlaneBasePointCoords, a_mat);
+        bool converged = MathUtils<double>::EigenSystem<3>(a_mat, eigenvector_mat, eigenval_mat);
+        KRATOS_ERROR_IF(!converged) << "Plane normal can't be computed. Eigenvalue problem did not converge." << std::endl;
+
+        // Find the minimum eigenvalue
+        double min_eigval = 0.0;
```
Comment (pooyan-dadvand) [inline kratos/utilities/plane_approximation_utility.h:200]: Are you sure that min_eigval is always negative? It shouldn't be initialized with std::numerical_limits::max?

File: kratos/utilities/plane_approximation_utility.h [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        unsigned int min_eigval_id = 0;
+        for (unsigned int i = 0; i < 3; ++i){
+            if (eigenval_mat(i,i) < min_eigval){
+                min_eigval_id = i;
+                min_eigval = std::min(min_eigval, eigenval_mat(i,i));
```
Comment (pooyan-dadvand) [inline kratos/utilities/plane_approximation_utility.h:205]: why you need the min again?

File: kratos/utilities/plane_approximation_utility.h [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+     * @return rPlaneBasePointCoords Plane base point
+     * @return rPlaneNormal Plane normal
+     */
+    static void ComputePlaneApproximation(
+        const std::vector< array_1d<double,3> > &rPointsCoords,
+        array_1d<double,3> &rPlaneBasePointCoords,
+        array_1d<double,3> &rPlaneNormal)
+    {
+        GetPlaneBasePoint(rPointsCoords,rPlaneBasePointCoords);
+        GetPlaneNormal(rPointsCoords, rPlaneBasePointCoords, rPlaneNormal);
```
Comment (pooyan-dadvand) [inline kratos/utilities/plane_approximation_utility.h:100]: Calculate would be better than Get for naming this methods to give a better impression of their cost. (Get gives the impression of returning an stored value)

**Review by pooyan-dadvand — COMMENTED** [review]
I know that you continue work on this, so please consider my comments on the next PR

File: kratos/utilities/plane_approximation_utility.h (hunk 4 above)
Comment (rubenzorrilla) [reply kratos/utilities/plane_approximation_utility.h:100]: Ok. Done!.

File: kratos/utilities/plane_approximation_utility.h (hunk 3 above)
Comment (rubenzorrilla) [reply kratos/utilities/plane_approximation_utility.h:205]: My bad, it is completely useless.

File: kratos/utilities/plane_approximation_utility.h (hunk 2 above)
Comment (rubenzorrilla) [reply kratos/utilities/plane_approximation_utility.h:200]: You're right, I'm changing it.

Comment (rubenzorrilla) [reply kratos/utilities/plane_approximation_utility.h:200]: Indeed, this yields an error in 2D cases... Thanks a lot @pooyan-dadvand .

**Pull request merged**